findAll will fail if collection is not created

Steps to reproduce:

1. Run a sample code
2. Configure application.properties
3. Remove the collection will be used from Azure portal
4. Run findAll
5. Fail with below exception
```
Caused by: java.lang.IllegalStateException: com.microsoft.azure.documentdb.DocumentClientException: Message: {"Errors":["Resource Not Found"]}
......
......
, SDK: Microsoft.Azure.Documents.Common/2.0.0.0, StatusCode: NotFound
	at com.microsoft.azure.documentdb.internal.ErrorUtils.maybeThrowException(ErrorUtils.java:69)
	at com.microsoft.azure.documentdb.internal.GatewayProxy.performGetRequest(GatewayProxy.java:240)
	at com.microsoft.azure.documentdb.internal.GatewayProxy.doRead(GatewayProxy.java:89)
	at com.microsoft.azure.documentdb.internal.GatewayProxy.processMessage(GatewayProxy.java:336)
	at com.microsoft.azure.documentdb.DocumentClient$8.apply(DocumentClient.java:2985)
	at com.microsoft.azure.documentdb.internal.RetryUtility.executeDocumentClientRequest(RetryUtility.java:58)
	at com.microsoft.azure.documentdb.DocumentClient.doRead(DocumentClient.java:2991)
	at com.microsoft.azure.documentdb.DocumentClient.readCollection(DocumentClient.java:658)
	at com.microsoft.azure.documentdb.internal.routing.ClientCollectionCache.readCollection(ClientCollectionCache.java:33)
	... 7 more
```

--- a/src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbOperations.java
+++ b/src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbOperations.java
@@ -20,8 +20,7 @@ public interface DocumentDbOperations {
 
     String getCollectionName(Class<?> entityClass);
 
-    DocumentCollection createCollectionIfNotExists(DocumentDbEntityInformation information,
-                                                   String partitionKeyFieldName);
+    DocumentCollection createCollectionIfNotExists(DocumentDbEntityInformation information);
 
     <T> List<T> findAll(Class<T> entityClass);
 

--- a/src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java
+++ b/src/main/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplate.java
@@ -290,13 +290,13 @@ public class DocumentDbTemplate implements DocumentDbOperations, ApplicationCont
     }
 
     @Override
-    public DocumentCollection createCollectionIfNotExists(@NonNull DocumentDbEntityInformation information,
-                                                          String partitionKeyFieldName) {
+    public DocumentCollection createCollectionIfNotExists(@NonNull DocumentDbEntityInformation information) {
         if (this.databaseCache == null) {
             this.databaseCache = createDatabaseIfNotExists(this.databaseName);
         }
 
         final String collectionName = information.getCollectionName();
+        final String partitionKeyFieldName = information.getPartitionKeyFieldName();
 
         final List<DocumentCollection> collectionList = getDocumentClient()
                 .queryCollections(getDatabaseLink(this.databaseName),

--- a/src/main/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/SimpleDocumentDbRepository.java
+++ b/src/main/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/SimpleDocumentDbRepository.java
@@ -7,6 +7,7 @@
 package com.microsoft.azure.spring.data.cosmosdb.repository.support;
 
 
+import com.microsoft.azure.documentdb.DocumentCollection;
 import com.microsoft.azure.documentdb.PartitionKey;
 import com.microsoft.azure.spring.data.cosmosdb.core.DocumentDbOperations;
 import com.microsoft.azure.spring.data.cosmosdb.core.query.Criteria;
@@ -28,19 +29,33 @@ import java.util.Optional;
 
 public class SimpleDocumentDbRepository<T, ID extends Serializable> implements DocumentDbRepository<T, ID> {
 
-    private final DocumentDbOperations documentDbOperations;
+    private final DocumentDbOperations operation;
     private final DocumentDbEntityInformation<T, ID> information;
+    private final DocumentCollection collection;
 
     public SimpleDocumentDbRepository(DocumentDbEntityInformation<T, ID> metadata,
                                       ApplicationContext applicationContext) {
-        this.documentDbOperations = applicationContext.getBean(DocumentDbOperations.class);
+        this.operation = applicationContext.getBean(DocumentDbOperations.class);
         this.information = metadata;
+
+        collection = createCollectionIfNotExists();
     }
 
     public SimpleDocumentDbRepository(DocumentDbEntityInformation<T, ID> metadata,
                                       DocumentDbOperations dbOperations) {
-        this.documentDbOperations = dbOperations;
+        this.operation = dbOperations;
         this.information = metadata;
+
+        collection = createCollectionIfNotExists();
+    }
+
+    private DocumentCollection createCollectionIfNotExists() {
+        return this.operation.createCollectionIfNotExists(this.information);
+    }
+
+    @Override
+    public void deleteCollection() {
+        this.operation.deleteCollection(this.information.getCollectionName());
     }
 
     /**
@@ -54,17 +69,13 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public <S extends T> S save(S entity) {
         Assert.notNull(entity, "entity must not be null");
 
-        // create collection if not exists
-        documentDbOperations.createCollectionIfNotExists(this.information,
-                this.information.getPartitionKeyFieldName());
-
         // save entity
         if (information.isNew(entity)) {
-            return documentDbOperations.insert(information.getCollectionName(),
+            return operation.insert(information.getCollectionName(),
                     entity,
                     createKey(information.getPartitionKeyFieldValue(entity)));
         } else {
-            documentDbOperations.upsert(information.getCollectionName(),
+            operation.upsert(information.getCollectionName(),
                     entity, createKey(information.getPartitionKeyFieldValue(entity)));
         }
 
@@ -90,13 +101,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public <S extends T> Iterable<S> saveAll(Iterable<S> entities) {
         Assert.notNull(entities, "Iterable entities should not be null");
 
-        // create collection if not exists
-        documentDbOperations.createCollectionIfNotExists(this.information,
-                this.information.getPartitionKeyFieldName());
-
-        for (final S entity : entities) {
-            save(entity);
-        }
+        entities.forEach(this::save);
 
         return entities;
     }
@@ -108,7 +113,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
      */
     @Override
     public Iterable<T> findAll() {
-        return documentDbOperations.findAll(information.getCollectionName(), information.getJavaType());
+        return operation.findAll(information.getCollectionName(), information.getJavaType());
     }
 
     /**
@@ -121,15 +126,10 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public List<T> findAllById(Iterable<ID> ids) {
         Assert.notNull(ids, "Iterable ids should not be null");
 
-        final List<T> entities = new ArrayList<T>();
+        final List<T> entities = new ArrayList<>();
 
-        for (final ID id : ids) {
-            final Optional<T> entity = findById(id);
+        ids.forEach(id -> findById(id).ifPresent(entities::add));
 
-            if (entity.isPresent()) {
-                entities.add(entity.get());
-            }
-        }
         return entities;
     }
 
@@ -147,8 +147,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
             return Optional.empty();
         }
 
-        final T result = documentDbOperations.findById(
-                information.getCollectionName(), id, information.getJavaType());
+        final T result = operation.findById(information.getCollectionName(), id, information.getJavaType());
 
         return result == null ? Optional.empty() : Optional.of(result);
     }
@@ -160,7 +159,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
      */
     @Override
     public long count() {
-        return documentDbOperations.count(information.getCollectionName());
+        return operation.count(information.getCollectionName());
     }
 
     /**
@@ -172,7 +171,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public void deleteById(ID id) {
         Assert.notNull(id, "id to be deleted should not be null");
 
-        documentDbOperations.deleteById(information.getCollectionName(), id, null);
+        operation.deleteById(information.getCollectionName(), id, null);
     }
 
     /**
@@ -184,11 +183,11 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public void delete(T entity) {
         Assert.notNull(entity, "entity to be deleted should not be null");
 
-        final String paritionKeyValue = information.getPartitionKeyFieldValue(entity);
+        final String partitionKeyValue = information.getPartitionKeyFieldValue(entity);
 
-        documentDbOperations.deleteById(information.getCollectionName(),
+        operation.deleteById(information.getCollectionName(),
                 information.getId(entity),
-                paritionKeyValue == null ? null : new PartitionKey(paritionKeyValue));
+                partitionKeyValue == null ? null : new PartitionKey(partitionKeyValue));
     }
 
     /**
@@ -196,7 +195,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
      */
     @Override
     public void deleteAll() {
-        documentDbOperations.deleteAll(information.getCollectionName(), information.getJavaType());
+        operation.deleteAll(information.getCollectionName(), information.getJavaType());
     }
 
     /**
@@ -208,9 +207,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public void deleteAll(Iterable<? extends T> entities) {
         Assert.notNull(entities, "Iterable entities should not be null");
 
-        for (final T entity : entities) {
-            delete(entity);
-        }
+        entities.forEach(this::delete);
     }
 
     /**
@@ -237,7 +234,7 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
         Assert.notNull(sort, "sort of findAll should not be null");
         final DocumentQuery query = new DocumentQuery(Criteria.getInstance(CriteriaType.ALL)).with(sort);
 
-        return documentDbOperations.find(query, information.getJavaType(), information.getCollectionName());
+        return operation.find(query, information.getJavaType(), information.getCollectionName());
     }
 
     /**
@@ -251,6 +248,6 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
     public Page<T> findAll(Pageable pageable) {
         Assert.notNull(pageable, "pageable should not be null");
 
-        return documentDbOperations.findAll(pageable, information.getJavaType(), information.getCollectionName());
+        return operation.findAll(pageable, information.getJavaType(), information.getCollectionName());
     }
 }

--- a/src/main/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/SimpleDocumentDbRepository.java
+++ b/src/main/java/com/microsoft/azure/spring/data/cosmosdb/repository/support/SimpleDocumentDbRepository.java
@@ -53,11 +53,6 @@ public class SimpleDocumentDbRepository<T, ID extends Serializable> implements D
         return this.operation.createCollectionIfNotExists(this.information);
     }
 
-    @Override
-    public void deleteCollection() {
-        this.operation.deleteCollection(this.information.getCollectionName());
-    }
-
     /**
      * save entity without partition
      *

--- a/src/test/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplateIT.java
+++ b/src/test/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplateIT.java
@@ -88,7 +88,7 @@ public class DocumentDbTemplateIT {
         dbConverter = new MappingDocumentDbConverter(mappingContext, objectMapper);
         dbTemplate = new DocumentDbTemplate(dbFactory, dbConverter, DB_NAME);
 
-        collectionPerson = dbTemplate.createCollectionIfNotExists(this.personInfo, null);
+        collectionPerson = dbTemplate.createCollectionIfNotExists(this.personInfo);
         dbTemplate.insert(Person.class.getSimpleName(), TEST_PERSON, null);
     }
 

--- a/src/test/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplatePartitionIT.java
+++ b/src/test/java/com/microsoft/azure/spring/data/cosmosdb/core/DocumentDbTemplatePartitionIT.java
@@ -9,7 +9,6 @@ package com.microsoft.azure.spring.data.cosmosdb.core;
 import com.fasterxml.jackson.databind.ObjectMapper;
 import com.microsoft.azure.documentdb.PartitionKey;
 import com.microsoft.azure.spring.data.cosmosdb.DocumentDbFactory;
-import com.microsoft.azure.spring.data.cosmosdb.common.TestConstants;
 import com.microsoft.azure.spring.data.cosmosdb.config.DocumentDBConfig;
 import com.microsoft.azure.spring.data.cosmosdb.core.convert.MappingDocumentDbConverter;
 import com.microsoft.azure.spring.data.cosmosdb.core.mapping.DocumentDbMappingContext;
@@ -17,7 +16,7 @@ import com.microsoft.azure.spring.data.cosmosdb.core.query.Criteria;
 import com.microsoft.azure.spring.data.cosmosdb.core.query.CriteriaType;
 import com.microsoft.azure.spring.data.cosmosdb.core.query.DocumentDbPageRequest;
 import com.microsoft.azure.spring.data.cosmosdb.core.query.DocumentQuery;
-import com.microsoft.azure.spring.data.cosmosdb.domain.Person;
+import com.microsoft.azure.spring.data.cosmosdb.domain.PartitionPerson;
 import com.microsoft.azure.spring.data.cosmosdb.repository.support.DocumentDbEntityInformation;
 import org.junit.After;
 import org.junit.Before;
@@ -48,11 +47,11 @@ import static org.junit.Assert.assertTrue;
 @RunWith(SpringJUnit4ClassRunner.class)
 @PropertySource(value = {"classpath:application.properties"})
 public class DocumentDbTemplatePartitionIT {
-    private static final Person TEST_PERSON = new Person(TestConstants.ID_1, TestConstants.FIRST_NAME,
-            TestConstants.LAST_NAME, TestConstants.HOBBIES, TestConstants.ADDRESSES);
+    private static final PartitionPerson TEST_PERSON = new PartitionPerson(ID_1, FIRST_NAME, LAST_NAME,
+            HOBBIES, ADDRESSES);
 
-    private static final Person TEST_PERSON_2 = new Person(TestConstants.ID_2, TestConstants.NEW_FIRST_NAME,
-            TEST_PERSON.getLastName(), TestConstants.HOBBIES, TestConstants.ADDRESSES);
+    private static final PartitionPerson TEST_PERSON_2 = new PartitionPerson(ID_2, NEW_FIRST_NAME,
+            TEST_PERSON.getLastName(), HOBBIES, ADDRESSES);
 
     @Value("${cosmosdb.uri}")
     private String documentDbUri;
@@ -60,11 +59,8 @@ public class DocumentDbTemplatePartitionIT {
     private String documentDbKey;
 
     private DocumentDbTemplate dbTemplate;
-    private MappingDocumentDbConverter dbConverter;
-    private DocumentDbMappingContext mappingContext;
-    private ObjectMapper objectMapper;
-    private DocumentDbEntityInformation<Person, String> personInfo;
     private String collectionName;
+    private DocumentDbEntityInformation<PartitionPerson, String> personInfo;
 
     @Autowired
     private ApplicationContext applicationContext;
@@ -73,38 +69,40 @@ public class DocumentDbTemplatePartitionIT {
     public void setup() throws ClassNotFoundException {
         final DocumentDBConfig dbConfig = DocumentDBConfig.builder(documentDbUri, documentDbKey, DB_NAME).build();
         final DocumentDbFactory dbFactory = new DocumentDbFactory(dbConfig);
+        final ObjectMapper objectMapper = new ObjectMapper();
+        final DocumentDbMappingContext mappingContext = new DocumentDbMappingContext();
 
-        mappingContext = new DocumentDbMappingContext();
-        objectMapper = new ObjectMapper();
+        personInfo = new DocumentDbEntityInformation<>(PartitionPerson.class);
         mappingContext.setInitialEntitySet(new EntityScanner(this.applicationContext).scan(Persistent.class));
 
-        dbConverter = new MappingDocumentDbConverter(mappingContext, objectMapper);
+        final MappingDocumentDbConverter dbConverter = new MappingDocumentDbConverter(mappingContext, objectMapper);
 
-        dbTemplate = new DocumentDbTemplate(dbFactory, dbConverter, TestConstants.DB_NAME);
-        personInfo = new DocumentDbEntityInformation<>(Person.class);
+        dbTemplate = new DocumentDbTemplate(dbFactory, dbConverter, DB_NAME);
         collectionName = personInfo.getCollectionName();
 
-        dbTemplate.createCollectionIfNotExists(this.personInfo, PROPERTY_LAST_NAME);
-        dbTemplate.insert(Person.class.getSimpleName(), TEST_PERSON, new PartitionKey(TEST_PERSON.getLastName()));
+        dbTemplate.createCollectionIfNotExists(personInfo);
+        dbTemplate.insert(PartitionPerson.class.getSimpleName(), TEST_PERSON,
+                new PartitionKey(TEST_PERSON.getLastName()));
     }
 
     @After
     public void cleanup() {
-        dbTemplate.deleteCollection(Person.class.getSimpleName());
+        dbTemplate.deleteAll(personInfo.getCollectionName(), PartitionPerson.class);
     }
 
     @Test
     public void testFindWithPartition() {
         Criteria criteria = Criteria.getInstance(IS_EQUAL, PROPERTY_LAST_NAME, Arrays.asList(LAST_NAME));
         DocumentQuery query = new DocumentQuery(criteria);
-        List<Person> result = dbTemplate.find(query, Person.class, Person.class.getSimpleName());
+        List<PartitionPerson> result = dbTemplate.find(query, PartitionPerson.class,
+                PartitionPerson.class.getSimpleName());
 
         assertThat(result.size()).isEqualTo(1);
         assertEquals(result.get(0), TEST_PERSON);
 
         criteria = Criteria.getInstance(IS_EQUAL, PROPERTY_ID, Arrays.asList(ID_1));
         query = new DocumentQuery(criteria);
-        result = dbTemplate.find(query, Person.class, Person.class.getSimpleName());
+        result = dbTemplate.find(query, PartitionPerson.class, PartitionPerson.class.getSimpleName());
 
         assertThat(result.size()).isEqualTo(1);
         assertEquals(result.get(0), TEST_PERSON);
@@ -115,35 +113,37 @@ public class DocumentDbTemplatePartitionIT {
         final Criteria criteria = Criteria.getInstance(IS_EQUAL, PROPERTY_ID, Arrays.asList(NOT_EXIST_ID));
         final DocumentQuery query = new DocumentQuery(criteria);
 
-        final List<Person> result = dbTemplate.find(query, Person.class, Person.class.getSimpleName());
+        final List<PartitionPerson> result = dbTemplate.find(query, PartitionPerson.class,
+                PartitionPerson.class.getSimpleName());
         assertThat(result.size()).isEqualTo(0);
     }
 
     @Test
     public void testUpsertNewDocumentPartition() {
-        final String firstName = TestConstants.NEW_FIRST_NAME + "_" + UUID.randomUUID().toString();
-        final Person newPerson = new Person(null, firstName, TestConstants.NEW_LAST_NAME, null, null);
+        final String firstName = NEW_FIRST_NAME + "_" + UUID.randomUUID().toString();
+        final PartitionPerson newPerson = new PartitionPerson(null, firstName, NEW_LAST_NAME, null, null);
 
         final String partitionKeyValue = newPerson.getLastName();
-        dbTemplate.upsert(Person.class.getSimpleName(), newPerson, new PartitionKey(partitionKeyValue));
+        dbTemplate.upsert(PartitionPerson.class.getSimpleName(), newPerson, new PartitionKey(partitionKeyValue));
 
-        final List<Person> result = dbTemplate.findAll(Person.class);
+        final List<PartitionPerson> result = dbTemplate.findAll(PartitionPerson.class);
 
         assertThat(result.size()).isEqualTo(2);
 
-        final Person person = result.stream()
+        final PartitionPerson person = result.stream()
                 .filter(p -> p.getLastName().equals(partitionKeyValue)).findFirst().get();
         assertThat(person.getFirstName()).isEqualTo(firstName);
     }
 
     @Test
     public void testUpdatePartition() {
-        final Person updated = new Person(TEST_PERSON.getId(), TestConstants.UPDATED_FIRST_NAME,
+        final PartitionPerson updated = new PartitionPerson(TEST_PERSON.getId(), UPDATED_FIRST_NAME,
                 TEST_PERSON.getLastName(), TEST_PERSON.getHobbies(), TEST_PERSON.getShippingAddresses());
-        dbTemplate.upsert(Person.class.getSimpleName(), updated, new PartitionKey(updated.getLastName()));
+        dbTemplate.upsert(PartitionPerson.class.getSimpleName(), updated, new PartitionKey(updated.getLastName()));
 
-        final List<Person> result = dbTemplate.findAll(Person.class);
-        final Person person = result.stream().filter(p -> TEST_PERSON.getId().equals(p.getId())).findFirst().get();
+        final List<PartitionPerson> result = dbTemplate.findAll(PartitionPerson.class);
+        final PartitionPerson person = result.stream().filter(
+                p -> TEST_PERSON.getId().equals(p.getId())).findFirst().get();
 
         assertTrue(person.equals(updated));
     }
@@ -153,15 +153,15 @@ public class DocumentDbTemplatePartitionIT {
         // insert new document with same partition key
         dbTemplate.insert(TEST_PERSON_2, new PartitionKey(TEST_PERSON_2.getLastName()));
 
-        final List<Person> inserted = dbTemplate.findAll(Person.class);
+        final List<PartitionPerson> inserted = dbTemplate.findAll(PartitionPerson.class);
         assertThat(inserted.size()).isEqualTo(2);
         assertThat(inserted.get(0).getLastName()).isEqualTo(TEST_PERSON.getLastName());
         assertThat(inserted.get(1).getLastName()).isEqualTo(TEST_PERSON.getLastName());
 
-        dbTemplate.deleteById(Person.class.getSimpleName(),
+        dbTemplate.deleteById(PartitionPerson.class.getSimpleName(),
                 TEST_PERSON.getId(), new PartitionKey(TEST_PERSON.getLastName()));
 
-        final List<Person> result = dbTemplate.findAll(Person.class);
+        final List<PartitionPerson> result = dbTemplate.findAll(PartitionPerson.class);
         assertThat(result.size()).isEqualTo(1);
         assertTrue(result.get(0).equals(TEST_PERSON_2));
     }
@@ -185,7 +185,7 @@ public class DocumentDbTemplatePartitionIT {
                 Arrays.asList(TEST_PERSON_2.getFirstName()));
         final DocumentQuery query = new DocumentQuery(criteria);
 
-        final long count = dbTemplate.count(query, Person.class, collectionName);
+        final long count = dbTemplate.count(query, PartitionPerson.class, collectionName);
         assertThat(count).isEqualTo(1);
     }
 
@@ -195,7 +195,7 @@ public class DocumentDbTemplatePartitionIT {
                 Arrays.asList("non-exist-first-name"));
         final DocumentQuery query = new DocumentQuery(criteria);
 
-        final long count = dbTemplate.count(query, Person.class, collectionName);
+        final long count = dbTemplate.count(query, PartitionPerson.class, collectionName);
         assertThat(count).isEqualTo(0);
     }
 
@@ -204,12 +204,13 @@ public class DocumentDbTemplatePartitionIT {
         dbTemplate.insert(TEST_PERSON_2, new PartitionKey(TEST_PERSON_2.getLastName()));
 
         final DocumentDbPageRequest pageRequest = new DocumentDbPageRequest(0, PAGE_SIZE_1, null);
-        final Page<Person> page1 = dbTemplate.findAll(pageRequest, Person.class, collectionName);
+        final Page<PartitionPerson> page1 = dbTemplate.findAll(pageRequest, PartitionPerson.class, collectionName);
 
         assertThat(page1.getContent().size()).isEqualTo(PAGE_SIZE_1);
         validateNonLastPage(page1, PAGE_SIZE_1);
 
-        final Page<Person> page2 = dbTemplate.findAll(page1.getPageable(), Person.class, collectionName);
+        final Page<PartitionPerson> page2 = dbTemplate.findAll(page1.getPageable(),
+                PartitionPerson.class, collectionName);
         assertThat(page2.getContent().size()).isEqualTo(1);
         validateLastPage(page2, PAGE_SIZE_1);
     }
@@ -219,11 +220,11 @@ public class DocumentDbTemplatePartitionIT {
         dbTemplate.insert(TEST_PERSON_2, new PartitionKey(TEST_PERSON_2.getLastName()));
 
         final Criteria criteria = Criteria.getInstance(CriteriaType.IS_EQUAL, "firstName",
-                Arrays.asList(TestConstants.FIRST_NAME));
+                Arrays.asList(FIRST_NAME));
         final PageRequest pageRequest = new DocumentDbPageRequest(0, PAGE_SIZE_2, null);
         final DocumentQuery query = new DocumentQuery(criteria).with(pageRequest);
 
-        final Page<Person> page = dbTemplate.paginationQuery(query, Person.class, collectionName);
+        final Page<PartitionPerson> page = dbTemplate.paginationQuery(query, PartitionPerson.class, collectionName);
         assertThat(page.getContent().size()).isEqualTo(1);
         validateLastPage(page, PAGE_SIZE_2);
     }

--- a/src/test/java/com/microsoft/azure/spring/data/cosmosdb/domain/PartitionPerson.java
+++ b/src/test/java/com/microsoft/azure/spring/data/cosmosdb/domain/PartitionPerson.java
@@ -1,0 +1,29 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+
+package com.microsoft.azure.spring.data.cosmosdb.domain;
+
+import com.microsoft.azure.spring.data.cosmosdb.core.mapping.PartitionKey;
+import lombok.AllArgsConstructor;
+import lombok.Data;
+
+import java.util.List;
+
+@Data
+@AllArgsConstructor
+public class PartitionPerson {
+
+    private String id;
+
+    private String firstName;
+
+    @PartitionKey
+    private String lastName;
+
+    private List<String> hobbies;
+
+    private List<Address> shippingAddresses;
+}

--- a/src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/integration/DocumentDBAnnotationIT.java
+++ b/src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/integration/DocumentDBAnnotationIT.java
@@ -74,8 +74,8 @@ public class DocumentDBAnnotationIT {
         dbClient = new DocumentClient(dbUri, dbKey, ConnectionPolicy.GetDefault(), ConsistencyLevel.Session);
         dbTemplate = new DocumentDbTemplate(dbFactory, mappingConverter, TestConstants.DB_NAME);
 
-        collectionRole = dbTemplate.createCollectionIfNotExists(roleInfo, null);
-        collectionExample = dbTemplate.createCollectionIfNotExists(sampleInfo, null);
+        collectionRole = dbTemplate.createCollectionIfNotExists(roleInfo);
+        collectionExample = dbTemplate.createCollectionIfNotExists(sampleInfo);
 
         dbTemplate.insert(roleInfo.getCollectionName(), TEST_ROLE, null);
     }

--- a/src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/integration/RoleRepositoryCollectionIT.java
+++ b/src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/integration/RoleRepositoryCollectionIT.java
@@ -1,0 +1,91 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+package com.microsoft.azure.spring.data.cosmosdb.repository.integration;
+
+import com.microsoft.azure.spring.data.cosmosdb.core.query.DocumentDbPageRequest;
+import com.microsoft.azure.spring.data.cosmosdb.domain.IntegerIdDomain;
+import com.microsoft.azure.spring.data.cosmosdb.exception.DocumentDBAccessException;
+import com.microsoft.azure.spring.data.cosmosdb.repository.TestRepositoryConfig;
+import com.microsoft.azure.spring.data.cosmosdb.repository.repository.IntegerIdDomainRepository;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.data.domain.Sort;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.SpringJUnit4ClassRunner;
+
+import java.util.Collections;
+
+@RunWith(SpringJUnit4ClassRunner.class)
+@ContextConfiguration(classes = TestRepositoryConfig.class)
+public class RoleRepositoryCollectionIT {
+
+    private static final IntegerIdDomain DOMAIN = new IntegerIdDomain(2, "fake-name");
+
+    @Autowired
+    private IntegerIdDomainRepository repository;
+
+    @Test
+    public void testBasicQuery() {
+        this.repository.save(DOMAIN);
+    }
+
+    @Test
+    public void testSaveAll() {
+        this.repository.saveAll(Collections.singleton(DOMAIN));
+    }
+
+    @Test
+    public void testFindAll() {
+        this.repository.findAll();
+    }
+
+    @Test
+    public void testFindAllById() {
+        this.repository.findAllById(Collections.singleton(DOMAIN.getNumber()));
+    }
+
+    @Test
+    public void testFindById() {
+        this.repository.findById(DOMAIN.getNumber());
+    }
+
+    @Test
+    public void testCount() {
+        this.repository.count();
+    }
+
+    @Test(expected = DocumentDBAccessException.class)
+    public void testDeleteById() {
+        this.repository.deleteById(DOMAIN.getNumber());
+    }
+
+    @Test(expected = DocumentDBAccessException.class)
+    public void testDelete() {
+        this.repository.delete(DOMAIN);
+    }
+
+    @Test
+    public void testDeleteAll() {
+        this.repository.deleteAll(Collections.singleton(DOMAIN));
+    }
+
+    @Test
+    public void testExistsById() {
+        this.repository.existsById(DOMAIN.getNumber());
+    }
+
+    @Test
+    public void testFindAllSort() {
+        this.repository.findAll(Sort.unsorted());
+    }
+
+    @Test
+    public void testFindAllPageable() {
+        final DocumentDbPageRequest pageRequest = new DocumentDbPageRequest(0, 3, null);
+        this.repository.findAll(pageRequest);
+    }
+}

--- a/src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/repository/PartitionPersonRepository.java
+++ b/src/test/java/com/microsoft/azure/spring/data/cosmosdb/repository/repository/PartitionPersonRepository.java
@@ -1,0 +1,14 @@
+/**
+ * Copyright (c) Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE in the project root for
+ * license information.
+ */
+package com.microsoft.azure.spring.data.cosmosdb.repository.repository;
+
+import com.microsoft.azure.spring.data.cosmosdb.domain.PartitionPerson;
+import com.microsoft.azure.spring.data.cosmosdb.repository.DocumentDbRepository;
+import org.springframework.stereotype.Repository;
+
+@Repository
+public interface PartitionPersonRepository extends DocumentDbRepository<PartitionPerson, String> {
+}

